lib/connection.js

```diff
@@ -1248,23 +1252,60 @@ Connection.prototype.batchDescribe = Connection.prototype.batchDescribeSObjects
   var self = this;
   var types = options.types;
   var autofetch = options.autofetch || false;
+  var maxConcurrentRequests = Math.min((options.maxConcurrentRequests || 15), 15);
   var batches = [];
   do {
     var batch = types.length > MAX_BATCH_REQUESTS ? types.slice(0, MAX_BATCH_REQUESTS) : types;
     batches.push(batch);
     types = types.length > MAX_BATCH_REQUESTS ? types.slice(MAX_BATCH_REQUESTS) : [];
   } while (types.length > 0 && autofetch);
-  return Promise.all(
-    batches.map(function (batch) { return self.doBatchDescribeRequest(batch); })
-    ).then(function (results) {
-      var sobjects = [];
-      results.forEach(function (sobjectArray) {
-        sobjectArray.forEach(function (sobject) { sobjects.push(sobject); })
-      });
-      return Promise.resolve(sobjects);
-    }).thenCall(callback);
+  var requestBatches = [];
+  do {
+    var requestBatch = batches.length > maxConcurrentRequests ? batches.slice(0, maxConcurrentRequests) : batches;
+    requestBatches.push(requestBatch);
+    batches = batches.length > maxConcurrentRequests ? batches.slice(maxConcurrentRequests) : [];
+  } while (batches.length > 0);
+  return self.doBatchDescribeRequestBatches(requestBatches)
+    .thenCall(callback);
 };
 
+Connection.prototype.doBatchDescribeRequestBatches = function(requestBatches) {
+  // make each batch of requests sequentially
+  var self = this;
+  var sobjects = [];
+  var firstBatch = requestBatches.shift();
+  return self.doBatchOfBatchDescribeRequests(firstBatch).then(
+    function (sobjectArray) {
+      sobjectArray.forEach(function (sobject) { sobjects.push(sobject); });
+      if (requestBatches.length > 0) {
+        return self.doBatchDescribeRequestBatches(requestBatches).then(
```

Comment:
This is kinda an interesting way to do the recursion. Couldn't you simplify this by just passing in the sobjects and condensing the functions?

```
Connection.prototype.doBatchDescribeRequestBatches = function(requestBatches, sobjects) {
  // make each batch of requests sequentially
  var self = this;
  var requestBatch = requestBatches.shift();

  if (!sobjects) sobjects = [];

  if (requestBatch) {
    return Promise.all(
      requestBatch.map(function (batch) { return self.doBatchDescribeRequest(batch); } )
    ).then(function (results) {
      results.forEach(function (sobjectArray) {
        sobjectArray.forEach(function (sobject) { sobjects.push(sobject); })
      });
      .then(function() {
        return this.doBatchDescribeRequestBatches(requestBatches, sobjects);
      });
  } else {
    return Promise.resolve(sobjects)
  }
}
```

Although, I haven't tested that and "simplified" in this case may be subjective.